### ISSUE TITLE
Add PlayerJumpEvent

### ISSUE THREAD

**Issue by Radviger** [SpongePowered/SpongeAPI#748]
Like in a forge. Just because it wasn't in Bukkit.

**Comment by JBYoshi** [comment]
I'm not sure if this is even possible to do. C0CPacketInput has a jump field, but it's only sent when a player is riding another entity (see EntityPlayerSP line 170). The jump() method is called when an entity jumps, but in the case of players, it's only on the client side.

**Comment by Guichaguri** [comment]
afaik this is 100% client-side, the client just send to the server the new location of the player, that's why fly hacks are pretty easy to make

**Comment by FerusGrim** [comment]
It's not exactly difficult to tell if someone is jumping.

Excuse the psuedo-code.

```java
@Subscribe
public void onPlayerMoveEvent(PlayerMoveEvent event) {
    Player player = event.getPlayer();
    if (player.isFlying() {
        return;
    }
    if (event.getOldLocation().getY() < event.getNewLocation().getY()) {
        // player is flying
    }
}
```

**Comment by ryantheleach** [comment]
Where does that +1 come from? do players always go up by 1 for every move event when jumping? seems awfully hacky and jittery.

**Comment by ST-DDT** [comment]
Rabbit/Jump Potions and other Jump Attribute Modifiers will crush that.
IIRC this was `OldY < NewY` before which would be more correct than this one.
However you have to exclude knockback, teleports, stairs/slabs...

**Comment by Xemiru** [comment]
Goes by the rule that we aren't able to assume anything. The jump can be mitigated by all sorts of effects, they can even be negated to less than a block. If that happens, that pseudo-code won't be able to fire and the jump event becomes unreliable. This isn't something we can implement because of how Mojang did things; the client has to tell us that they jumped.

The usual alternative I see to this however is checking if the player tried to fly like in creative mode. It's slightly different, in the sense that its looking for a doublejump as opposed to just a single jump, however if the client lags then they're able to fly around as they please, and end up where they flew unless its far enough for the server to fire a moved-too-quickly-dropkick.

tl;dr limitations pls

**Comment by intronate67** [comment]
Creating an event for this is as inaccurate as checking it in a plugin.
The client (right now, this might change with the modding-API) only sends the coordinates it moves to, and the location it faces to, not the actual button presses. This should be closed. It would be nice to be able to check it but it is impossible to.

**Comment by thecodewarrior** [comment]
I don't have a very in-depth knowledge of the minecraft server, but couldn't you check if the player moved up and isOnGround went from true to false?

**Comment by kenzierocks** [comment]
No, think about knockback.

**Comment by Xemiru** [comment]
@thecodewarrior It is only accurate client-side, on the server's side it can't be trusted because its the client that decides that information. That said, any mod can modify it so it always returns true or false.

**Comment by zml2008** [comment]
Yeah, closing because the client just doesn't send this information to the server, so it's not something that can be done in a serverside API.
